src/Symfony/Component/Security/Core/User/InMemoryUserProvider.php

```diff
@@ -90,6 +90,19 @@ public function supportsClass($class)
         return 'Symfony\Component\Security\Core\User\User' === $class;
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function upgradePassword(UserInterface $user, string $newEncodedPassword): void
+    {
+        if (!$user instanceof User) {
+            throw new UnsupportedUserException(sprintf('Instances of "%s" are not supported.', \get_class($user)));
+        }
+
+        $this->getUser($user->getUsername())->setPassword($newEncodedPassword);
+        $user->setPassword($newEncodedPassword);
```

Comment:
Shouldn't this just throw an exception? Changing the password of an in-memory doesn't make sense as it be reverted back to the hardcoded value on the next request.

Reply:
what about tests?

Reply:
Upgrades are opportunistic, so this is fine also for static credentials to me.

Reply:
> what about tests?

What about them? Having something in the production code just because you want to test it is putting the cart in front of the horse if you ask me.

My main gripe with this is that the API suggests now that the password of in-memory users can be changed, and junior developers with minimal experience might think this actually works. I don't want to imagine the rabbit hole they fall into when they find it isn't working as they expected ...

Reply:
Agreed, interface removed from the class.